src/ipc/ipc4/mtrace.c

```diff
@@ -0,0 +1,172 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2021 Intel Corporation. All rights reserved.
+//
+// Author: Kai Vehmanen <kai.vehmanen@linux.intel.com>
+
+#include <stdint.h>
+#include <sof/lib/uuid.h>
+#include <sof/ipc/common.h>
+#include <ipc4/base_fw.h>
+#include <ipc4/error_status.h>
+
+#ifdef CONFIG_LOG_BACKEND_ADSP_MTRACE
+
+#undef panic // SOF headers may redefine panic
+#include <zephyr/logging/log_backend.h>
+#include <zephyr/logging/log_backend_adsp_mtrace.h>
+#include <zephyr/logging/log.h>
+
+#include <sof/schedule/ll_schedule.h>
+#include <sof/schedule/schedule.h>
+#include <sof/schedule/task.h>
+
+LOG_MODULE_REGISTER(mtrace, CONFIG_SOF_LOG_LEVEL);
+
+/**
+ * If mtrace log buffer has less space than
+ * the threshold, notify host with BUFFER_STATUS IPC4.
+ */
+#define NOTIFY_BUFFER_STATUS_THRESHOLD 2048
+
+/**
+ * Default aging timer value. This time defines the maximum time to
+ * block IPC BUFFER STATUS notify messages. The notification is sent
+ * either due to amount of data available in the buffer, or by timeout
+ * in case logs are sent at slow rate.
+ */
+#define IPC4_MTRACE_NOTIFY_AGING_TIMER_MS 1000
+
+/**
+ * Smallest possible aging timer value
+ */
+#define IPC4_MTRACE_AGING_TIMER_MIN_MS 100
+
+/* bb2aa22e-1ab6-4650-8501-6e67fcc04f4e */
+DECLARE_SOF_UUID("mtrace-task", mtrace_task_uuid, 0xbb2aa22e, 0x1ab6, 0x4650,
+		 0x85, 0x01, 0x6e, 0x67, 0xfc, 0xc0, 0x4f, 0x4e);
+
+static uint64_t mtrace_notify_last_sent;
+
+static uint32_t mtrace_bytes_pending;
+
+static uint32_t mtrace_aging_timer = IPC4_MTRACE_NOTIFY_AGING_TIMER_MS;
+
+static struct k_thread mtrace_thread;
+
+#define MTRACE_STACK_SIZE	4096
+#define MTRACE_IPC_CORE	PLATFORM_PRIMARY_CORE_ID
+#define MTRACE_THREAD_PRIO      -1
+#define MTRACE_THREAD_NAME      "mtrace"
+
+K_KERNEL_STACK_DEFINE(mtrace_stack, MTRACE_STACK_SIZE);
+
+static void mtrace_log_hook(size_t written, size_t space_left)
+{
+	if (arch_proc_id() == MTRACE_IPC_CORE &&
+	    (space_left < NOTIFY_BUFFER_STATUS_THRESHOLD ||
+	     k_uptime_delta(&mtrace_notify_last_sent) >= mtrace_aging_timer)) {
```

Comment:
does this mean, that if core 1 logs a lot and core 0 doesn't log at all, we'll potentially lose those logs? Shouldn't we `k_wakeup()` if running on a different core and reaching the threshold?

Reply:
Yeah, the multicore part will require more work. The mtrace protocol actually defines separate slots of each core, but this is not part of this PR (plus with current windows set up by the bootloader, we don't have enough space for this approach). So this PR's approach to multicore is  a simple one. The logs won't be lost as the mtrace_thread (pinned on core0) will flush the logs for all cores. It does mean that in some multicore cases user-space has to set a lower aging timer value (this can be set via debugfs already now). 

In longer term, the current 4K slot is too little to relay logs for all cores, so we'll need a bigger rework, but that will be done in later PRs.

Reply:
> The logs won't be lost as the mtrace_thread (pinned on core0) will flush the logs for all cores.

But if one or several secondary cores log more than the buffer can hold between two thread runs, which are currently set to 1s? You can log a lot in 1s.

Reply:
@lyakh Kernel sets this by default to 250ms. It's still a lot of time, so I think the eventual solution is to use the per-core buffers.

Reply:
why would per-core buffers help? With 4 cores you'd be better off putting all the memory into one buffer than splitting it into 4 parts? But either way, doesn't increasing the buffer size just make this problem less likely to bite us instead of fixing it? The solution seems to be to make sure logging on secondary cores wakes up the logging thread if needed - therefore `k_wakeup()`

Reply:
The max buffer per core is 4K (as per protocol). But it is allowed to have a separate buffer for each core, so that increases total log buffering to 4x4K on a 4core system.